File: dfs_bfs_backtracking/python/b10026.py
N = int(input())
pictures = [input() for _ in range(N)]
visited = [[False for _ in range(N)] for __ in range(N)]
directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

def check(i, j, blind_checked):
    front = -1
    rear = 0
    queue = [0] * (N * N)

    queue[rear] = (i, j)
    visited[i][j] = not blind_checked

    cmpchar = pictures[i][j]

    while front != rear:
        front += 1
        row, col = queue[front]
        for direction in directions:
            new_row = row + direction[0]
            new_col = col + direction[1]
            add_flag = False
            if 0 <= new_row < N and 0 <= new_col < N and visited[new_row][new_col] == blind_checked:
                if blind_checked:
                    if cmpchar == 'B' and pictures[new_row][new_col] == 'B':
                        add_flag = True
                    if cmpchar != 'B' and pictures[new_row][new_col] != 'B':
                        add_flag = True
                else:
                    if cmpchar == pictures[new_row][new_col]:
                        add_flag = True
            if add_flag:
                visited[new_row][new_col] = not blind_checked
                rear += 1
                queue[rear] = (new_row, new_col)

group_num = group_blind_num = 0

for i in range(N):
    for j in range(N):
        if not visited[i][j]:
            check(i, j, False)
            group_num += 1

for i in range(N):
    for j in range(N):
        if  visited[i][j]:
            check(i, j, True)
            group_blind_num += 1

print(group_num, group_blind_num)

File: dfs_bfs_backtracking/cpp/b10026.cpp
#include <iostream>
#include <string>
#include <queue>
#include <vector>

int N;

std::string pictures[100];
bool visited[100][100]{ false };
std::vector<std::pair<int, int>> directions{ {1, 0}, {-1, 0}, {0, 1}, {0, -1} };


void check(const int& temp_row, const int& temp_col, const bool& blind_flag) {
	std::queue<std::pair<int, int>> check_queue;
	visited[temp_row][temp_col] = !blind_flag;
	check_queue.push(std::make_pair(temp_row, temp_col));
	char cmpchar = pictures[temp_row][temp_col];

	while (check_queue.size()) {
		int row, col;
		std::pair<int, int>& pix = check_queue.front();
		check_queue.pop();
		row = pix.first;
		col = pix.second;
		for (std::pair<int, int>& direction : directions) {
			int new_row = row + direction.first;
			int new_col = col + direction.second;
			if (new_row >= 0 && new_row < N && new_col >= 0 && new_col < N) {
				if (visited[new_row][new_col] == blind_flag) {
					if (blind_flag) {
						if (cmpchar == 'B' && pictures[new_row][new_col] == cmpchar) {
							visited[new_row][new_col] = !blind_flag;
							check_queue.push(std::make_pair(new_row, new_col));
						}
						if (cmpchar != 'B' && pictures[new_row][new_col] != 'B') {
							visited[new_row][new_col] = !blind_flag;
							check_queue.push(std::make_pair(new_row, new_col));
						}
					}
					else {
						if (pictures[new_row][new_col] == cmpchar) {
							visited[new_row][new_col] = !blind_flag;
							check_queue.push(std::make_pair(new_row, new_col));
						}
					}
				}
			}
		}
	}
}


int main() {
	std::cin >> N;
	for (int i = 0; i < N; ++i) {
		std::cin >> pictures[i];
	}

	int check_num = 0;
	int check_blind_num = 0;

	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < N; ++j) {
			if (visited[i][j] == 0) {
				check(i, j, false);
				check_num += 1;
			}
		}
	}

	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < N; ++j) {
			if (visited[i][j] == 1) {
				check(i, j, true);
				check_blind_num += 1;
			}
		}
	}
	std::cout << check_num << " " << check_blind_num << std::endl;
}